pkg/lib/cockpit-components-listing-panel.jsx

```diff
@@ -102,10 +105,12 @@ export class ListingPanel extends React.Component {
             if (tabIdx !== this.state.activeTab && !(tabIdx in this.state.loadedTabs))
                 continue;
             row = <Renderer key={ this.props.tabRenderers[tabIdx].name } hidden={ (tabIdx !== this.state.activeTab) } {...rendererData} />;
-            if (tabIdx === this.state.activeTab)
-                tabs.push(<div className="ct-listing-panel-body" key={tabIdx}>{row}</div>);
-            else
-                tabs.push(<div className="ct-listing-panel-body" key={tabIdx} hidden>{row}</div>);
+            if (!this.props.tabRenderers[tabIdx].hide) {
+                if (tabIdx === this.state.activeTab)
+                    tabs.push(<div className="ct-listing-panel-body" key={tabIdx}>{row}</div>);
+                else
+                    tabs.push(<div className="ct-listing-panel-body" key={tabIdx} hidden>{row}</div>);
+            }
```

Comment:
Shouldn't this push an empty render when hidden? Or is it meant to really entirely leave out the complete column when hidden?

What's the point of this "hide" flag then? Wouldn't it be cleaner to simply not render stuff at all?

Reply:
Oh so I had problem with this. Snapshots tab is rendered conditionally, only when user has snapshots APIs available on libvirt-dbus. To find this out, it takes some time. I could think of several solutions:
1. Do not pass snapshots tab to this component at all. This however causes a problem, because this component keeps track of which tab is rendered by activeTab property in state. That is just a simple index of position in array of tabs. So if if we find out that snapshots APIs are available, we have to insert snapshot tab into a list, which if user has opened any other tab (like consoles), it causes involuntary switch of tabs. I tried to update state when props change, but that is discouraged by React and following even official React solution `https://reactjs.org/docs/react-component.html#componentdidupdate` would cause syntax error (cannot call setState from ComponentDidUpdate even).
2. Using something else than index in array of tabs to track the number of the activeTab. Only id prop could be used, and that prop is not used in almost any other place this component is called from, so it could include rewriting a lot of code. I can do that if you think that's best solution.
3. Pass snapshot tab, but keep it hidden until cockpit finds out if user has snapshots APIs available. This would not disrupt problem with activeTab id.

So perhaps I'm missing out on something, not seeing the better solution. I'm open to discussion.

Reply:
I see -- then keep the `hide` approach for now, and let's just clean that up a bit. But it seems to me that with *this* solution you should get jumping tabs, as changing the "hide" value will *insert* a new element into the tab renderers, so the offset of all subsequent renderers changes.

Another option is to wait with rendering the entire properties tab until you find out whether the API is available. That doesn't take much time, and doesn't change over the lifetime of the page - i. e. the result can be cached and only needs to be determined once.

Reply:
I tried it and there will be no jumping tabs. Changing hide value won't insert a new element into an array. It's there, it just has hide property so it won't be rendered, but the index in a array of tabs will be consistent no matter what's the hide property.

Reply:
But alright, perhaps waiting with renderint the entire tabs until we know if snapshots apis are available is the best, since it's simplest solution :D